aidial_adapter_bedrock/llm/converse/adapter.py

```diff
@@ -0,0 +1,174 @@
+from aidial_adapter_bedrock.utils.list import omit_by_indices
+from aidial_adapter_bedrock.utils.list_projection import ListProjection
+
+
+class ConverseAdapter(ChatCompletionAdapter):
+    deployment: str
+    bedrock: Bedrock
+    storage: FileStorage | None
+
+    tokenize_text: Callable[[str], int] = default_tokenize_string
+    input_tokenizer_factory: Callable[
+        [ConverseDeployment, ConverseParams],
+        Callable[[List[Tuple[ConverseMessage, Any]]], Awaitable[int]],
+    ]
+    support_tools: bool
+    partitioner: Callable[[List[Tuple[ConverseMessage, Any]]], List[int]] = (
+        turn_based_partitioner
+    )
+
+    async def _discard_messages(
+        self, params: ConverseParams, max_prompt_tokens: int | None
+    ) -> Tuple[DiscardedMessages | None, ConverseParams]:
+        if max_prompt_tokens is None:
+            return None, params
+
+        discarded_messages, messages = await truncate_prompt(
+            messages=params.messages.list,
+            tokenizer=self.input_tokenizer_factory(self.deployment, params),
+            keep_message=keep_last,
+            partitioner=self.partitioner,
+            model_limit=None,
+            user_limit=max_prompt_tokens,
+        )
+
+        return list(
+            params.messages.to_original_indices(discarded_messages)
+        ), ConverseParams(
+            **{
+                **params.to_dict(),
+                "messages": ListProjection(
+                    omit_by_indices(messages, discarded_messages)
+                ),
+            },
+        )
+
+    async def count_prompt_tokens(
+        self, params: ModelParameters, messages: List[DialMessage]
+    ) -> int:
+        converse_params = await self.construct_converse_params(messages, params)
+        return await self.input_tokenizer_factory(
+            self.deployment, converse_params
+        )(converse_params.messages.list)
+
+    async def count_completion_tokens(self, string: str) -> int:
+        return self.tokenize_text(string)
+
+    async def compute_discarded_messages(
+        self, params: ModelParameters, messages: List[DialMessage]
+    ) -> DiscardedMessages | None:
+        converse_params = await self.construct_converse_params(messages, params)
+        discarded_messages, _ = await self._discard_messages(
+            converse_params, params.max_prompt_tokens
+        )
+        return discarded_messages
+
+    def get_tool_config(self, params: ModelParameters) -> ConverseTools | None:
+        if params.tool_config and not self.support_tools:
+            raise ValidationError("Tools are not supported")
+        return (
+            to_converse_tools(params.tool_config)
+            if params.tool_config
+            else None
+        )
+
+    async def construct_converse_params(
+        self,
+        messages: List[DialMessage],
+        params: ModelParameters,
+    ) -> ConverseParams:
+        system_message = get_converse_system_prompt(messages)
```

Comment:
Should this method also **remove** system message from the list of messages?

```python
system_message, messages = get_converse_system_prompt(messages)
```

Reply:
Done, https://github.com/epam/ai-dial-adapter-bedrock/pull/177#discussion_r1839843769